controllers/agents/app/servicebinding_controller.go

```diff
@@ -113,44 +151,61 @@ func (r *ServiceBindingReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 		return ctrl.Result{}, nil
 	}
 
-	applications, err := r.getApplication(ctx, req, serviceBinding, secretName)
-	if err != nil {
-		// error retrieving the application(s), so setting the service binding status to false and reconcile
-		if errUpdateStatus := r.setStatus(ctx, secretName, serviceBinding, metav1.ConditionFalse, conditionGetAppsFailureReason, primazaiov1alpha1.ServiceBindingStateMalformed, err.Error()); errUpdateStatus != nil {
-			return ctrl.Result{}, errUpdateStatus
+	l.Info("Add Finalizer if needed")
+	// add finalizer if needed
+	if !controllerutil.ContainsFinalizer(&serviceBinding, ServiceBindingFinalizer) {
+		controllerutil.AddFinalizer(&serviceBinding, ServiceBindingFinalizer)
+		if err := r.Update(ctx, &serviceBinding); err != nil {
+			return ctrl.Result{}, err
 		}
+	}
+
+	if err := r.SetWatchersForResources(ctx, serviceBinding); err != nil {
+		l.Error(err, "Failed to set watchers on ServiceBinding resources ", "namespace", req.Namespace, "name", req.Name)
+		return ctrl.Result{}, err
+	}
+
+	var psSecret v1.Secret
+	if psSecret, err = r.GetSecret(ctx, serviceBinding, applications); err != nil {
+		return ctrl.Result{}, err
+	}
+	if err = r.PrepareBinding(ctx, serviceBinding, applications, &psSecret); err != nil {
 		return ctrl.Result{}, err
 	}
 
+	if err := ctrl.SetControllerReference(&serviceBinding, &psSecret, r.Scheme); err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, r.Update(ctx, &psSecret)
+}
+
+func (r *ServiceBindingReconciler) GetSecret(ctx context.Context, serviceBinding v1alpha1.ServiceBinding, applications []unstructured.Unstructured) (v1.Secret, error) {
```

Comment:
I'd propose to use `*v1.Secret` as return type. WDYT?

```suggestion
func (r *ServiceBindingReconciler) GetSecret(ctx context.Context, serviceBinding v1alpha1.ServiceBinding, applications []unstructured.Unstructured) (*v1.Secret, error) {
```

Reply:
Makes sense